@yawkat there's probably something happening in the native stack of Netty which kills the communication between the Gradle daemon and the test process. However, you can run this single test: `./gradlew :http-server-netty:test --tests "io.micronaut.http.server.netty.binding.ByteBufferSpec.test reading the body with a publisher of bytebuffers"`

and see that it spits a lot of exceptions:

```
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, increment: 1
	at io.netty.util.internal.ReferenceCountUpdater.retain0(ReferenceCountUpdater.java:133)
	at io.netty.util.internal.ReferenceCountUpdater.retain(ReferenceCountUpdater.java:120)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.retain(AbstractReferenceCountedByteBuf.java:81)
	at io.netty.handler.codec.http.DefaultHttpContent.retain(DefaultHttpContent.java:68)
	at io.netty.handler.codec.http.DefaultHttpContent.retain(DefaultHttpContent.java:25)
	at io.netty.util.ReferenceCountUtil.retain(ReferenceCountUtil.java:41)
	at io.netty.handler.codec.http.HttpContentEncoder.encode(HttpContentEncoder.java:216)
	at io.micronaut.http.server.netty.SmartHttpContentCompressor.encode(SmartHttpContentCompressor.java:65)
	at io.micronaut.http.server.netty.SmartHttpContentCompressor.encode(SmartHttpContentCompressor.java:33)
	at io.netty.handler.codec.MessageToMessageCodec$1.encode(MessageToMessageCodec.java:67)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:90)
	... 47 common frames omitted
```

I got the stack trace from running the test in the IDE.

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java
@@ -573,9 +573,11 @@ public final class RoutingInBoundHandler implements RequestHandler {
             s.onSubscribe(new Subscription() {
                 @Override
                 public void request(long n) {
+                    HttpContent first = LazySendingSubscriber.this.first;
                     if (first != null) {
+                        LazySendingSubscriber.this.first = null;
+                        // onNext may trigger further request calls
                         s.onNext(first);
-                        first = null;
                         if (n != Long.MAX_VALUE) {
                             n--;
                             if (n == 0) {

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java
@@ -627,7 +627,11 @@ public final class RoutingInBoundHandler implements RequestHandler {
                 MutableHttpResponse<?> response;
                 if (t instanceof HttpStatusException hse) {
                     response = HttpResponse.status(hse.getStatus());
-                    hse.getBody().ifPresent(response::body);
+                    if (hse.getBody().isPresent()) {
+                        response.body(hse.getBody().get());
+                    } else if (hse.getMessage() != null) {
+                        response.body(hse.getMessage());
+                    }
                 } else {
                     response = routeExecutor.createDefaultErrorResponse(request, t);
                 }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
@@ -1,8 +1,6 @@
 package io.micronaut.http.server.netty.handler
 
-
 import io.micronaut.http.netty.stream.StreamedHttpRequest
-import io.micronaut.http.server.netty.DelegateStreamedHttpResponse
 import io.netty.buffer.Unpooled
 import io.netty.channel.ChannelHandlerContext
 import io.netty.channel.ChannelOutboundHandlerAdapter
@@ -81,7 +79,7 @@ class PipeliningServerHandlerSpec extends Specification {
         def ch = new EmbeddedChannel(mon, new PipeliningServerHandler(new RequestHandler() {
             @Override
             void accept(ChannelHandlerContext ctx, HttpRequest request, PipeliningServerHandler.OutboundAccess outboundAccess) {
-                outboundAccess.writeStreamed(new DelegateStreamedHttpResponse(resp, sink.asFlux()))
+                outboundAccess.writeStreamed(resp, sink.asFlux())
             }
 
             @Override

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/stream/ErrorResponseSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/stream/ErrorResponseSpec.groovy
@@ -29,7 +29,7 @@ class ErrorResponseSpec extends Specification {
         def e = thrown(HttpClientResponseException)
         verifyAll {
             e.status == HttpStatus.BAD_REQUEST
-            e.message == "Client '/flowable': expected flowable error"
+            e.response.getBody(String).get() == "expected flowable error"
         }
     }
 

